cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreManager.java

```diff
@@ -333,7 +333,7 @@ private Datastore createDatastore(String dbName, KeyProvider provider) throws Da
             // if it does not exist
 
             //Pass database directory, database name, and SQLCipher key provider
-            DatastoreImpl ds = new DatastoreImpl(dbDirectory, dbName, provider);
+            QueryableDatastore ds = new QueryableDatastore(dbDirectory, dbName, provider);
```

Comment:
I'm afraid I agree with @ricellis on his delegation point. I don't see how it makes sense for `DatastoreImpl` to be abstract (leaving naming conventions to one side). The problem of using inheritance is that the java model of single inheritance isn't the right tool for the job. It means that your class hierarchy is inverted - each sub-class is increasingly specialised whereas what we want is to add functionality with each layer (there is a subtle difference).

Consider that we might have other 'managers' with functionality analogous to Index/Query. We might have a geo-spatial manager which adds extra functionality, and we might also have (hypothetically) a large attachment manager which switches storage technologies so that attachments are stored somewhere other than Cloudant on the server side.

Depending on what capabilities the user wants, we now have:

```
DatastoreImpl
QueryableDatastore
GeospatialDatastore
LargeAttachmentDatastore
```

The user has to choose the one they want. But wait, we're not done, what if they want to combine the features?

```
QueryableGeospatialDatastore
QueryableLargeAttachmentDatastore
GeospatialLargeAttachmentDatastore
QueryableGeospatialLargeAttachmentDatastore
```

Well I think you can see where this is going. I honestly think that inheritance isn't the right tool for the job. Some sort of class composition as suggested by Rich would be better. There are a few options here, which maybe include the [Facade Pattern](https://en.wikipedia.org/wiki/Facade_pattern) (this is better explained in the _Design Patterns_ book).

The downside with the Facade Pattern is that we have to write a fair bit of boilerplate proxying calls to the other components ("sub-system classes", like `IndexManager`), but on the upside it has high usability for the API consumer. The alternative delegation method suggested like `getQueryableDatastore().find` requires a bit more typing by the API consumer but requires less boilerplate on our part.

I also think it might be worth prototyping some of these approaches before generating any larger diffs.

Comments welcome!